Error analysis pipeline

lizgzil opened the pull request:
Fixes #27



---

# Description

In this PR:
- A little refactor of some of Aidan's code [preweight_tenure_error_analysis.py](https://github.com/nestauk/asf_heat_pump_suitability/pull/28/files#diff-7f1d94c4e7a89b5e43b5b3bc0af3b3592f83652a5fa8f311c0a3a9aa904aba89)
- Run error analysis on all LSOAs for all features, before and after weighting
- Makes plots from the results

Fixes # (issue)

# Instructions for Reviewer

- Don't worry about `asf_heat_pump_suitability/getters/s3_getters.py` - it is all copied from previous projects!
- Aidan's already reviewed code has been split into `asf_heat_pump_suitability/pipeline/error_analysis/error_analysis_utils.py` and `asf_heat_pump_suitability/pipeline/error_analysis/sample/preweight_tenure_error_analysis.py`. the only thing that needs review in these is the `get_error_reduction` function in `error_analysis_utils`
- Please check for bugs etc in `asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py`
- Try running:
```
python asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py --sample
```

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

lizgzil committed:
Pull out Aidan's evaluation functions into a utils file, rename folder to not be specific to preweighting evaluation

lizgzil committed:
Refactor error function for efficiency in loop

lizgzil committed:
Add to requirements and add s3 getter functions

lizgzil committed:
Add the evaluation pipeline script

lizgzil committed:
Calculate error reduction in evaluation metrics

lizgzil committed:
add in rooms to the error analysis

lizgzil committed:
Add the analysis of the evaluation notebook

lizgzil committed:
Add notebook properly

crispy-wonton commented on asf_heat_pump_suitability/getters/s3_getters.py:
Some of these are missing docstrings and type hinting. They look really useful :)

crispy-wonton commented on asf_heat_pump_suitability/pipeline/error_analysis/error_analysis_utils.py:
```diff
@@ -185,14 +108,85 @@ def process_single_lsoa(
+    after_proportions: dict,
+    target_proportions: dict,
+) -> dict:
+    """
+    Find the average error reduction between two sets of proportions and the target
+
+    """
```
Missing args and returns. What are the keys and values in each dict? Is this used per feature or will one dict contain all the features?

Also, it would be really helpful to briefly explain what the output metric represents and the range of possible values (is it inf: -inf?). 

Please can you confirm if my interpretation is correct for my understanding:
- a negative error_reduction = weighting improved proportions closer to target
- positive error_reduction means weighting made the error worse
- more negative numbers mean there was greater improvement and lower remaining error
- negative numbers closer to zero mean there was less improvement and higher remaining error

Also, what happens if all the error was corrected and there is no remaining error? It looks like those instances aren't recorded - should they be?

lizgzil replied:
Great points of clarification. In considering these, I realised that the calculation is incorrect anyway.

I have changed it and gave a much more detailed docstring with the answers to the questions you've posed here.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/error_analysis/error_analysis_utils.py:
```diff
@@ -185,14 +108,85 @@ def process_single_lsoa(
+            ) / after_diff
+
+    if error_reduction:
+        return np.mean(list(error_reduction.values()))
```
since this function isnt returning the `error_reduction` dict, I think `error_reduction` can just be a list

lizgzil replied:
so true!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+
+[remove the --sample argument to run on full dataset]
+
+"""
```
This is helpful

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    "mae_missing_cats",
+]
+
+evaluation_feature_names = ["tenure", "property_type", "build_year", "nrooms"]
```
can these variables go in if name = main and be passed as args to functions

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    parser = ArgumentParser()
+
+    parser.add_argument(
+        "--reweighted_dir",
```
```suggestion
        "--reweighted_path",
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    Args:
+                    feature_name (str): The feature name to calculate errors for
+                    lsoa_code (str): The LSOA
+                    target_features (dict): A nested dictionary of the counts for all property features and LSOAs
+                    target_marginals (dict): A nested dictionary of the proportions for all property features and LSOAs
```
Does this mean input is like: 
`{"LSOA_code": {"feature1a": count, "feature1b": count, "feature2a": count}}` 
etc.?

Also is the only diff between target features and target marginals the fact that it's counts v proportions?

lizgzil replied:
not quite - it looks like `{'tenure': {'E01022833': {'owner-occupied': 548, 'rental (social)': 53}, 'E01013414': {}}, 'property_type': {}}` - I'll add it to the docstring.

Yes - thats right it is just the proportions vs counts

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    target_proportions = target_marginals[feature_name].get(lsoa_code)
+
+    if target_counts:
+        orig_counts_list = epc_subset[feature_name].value_counts().to_dict()
```
```suggestion
        orig_counts_dict = epc_subset[feature_name].value_counts().to_dict()
```
I got confused by the name on line 95 :D

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+
+from argparse import ArgumentParser
+
+from asf_heat_pump_suitability.pipeline.error_analysis.error_analysis_utils import *
```
I would import `error_analysis_utils` and then call functions from the module with `error_analysis_utils.function() `where required as I find it easier to locate origins of code that way. Same with modules below

lizgzil replied:
will do that! I don't know why I even did that because I also dislike `import *` for the same reason!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+        )
+        if not all(epc_subset["weight"].is_null()):
+            # Get the sum of the weights for each feature category (not technically a 'count')
+            reweighted_counts_list = (
```
```suggestion
            reweighted_counts_dict = (
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+            reweighted_counts_list = (
+                epc_subset.group_by(feature_name).agg(pl.col("weight").sum()).to_dict()
+            )
+            reweighted_counts = dict(
+                zip(
+                    reweighted_counts_list[feature_name],
+                    reweighted_counts_list["weight"],
+                )
+            )
```
Fyi, I think the counts will appear very wrong in the error_metrics_reweighted because this 'weight' col is based on an artificial population of 1000, so it's basically just the proportion * 1000. Did they look very off to you when you ran it? 

However, when we come to microsimulation, we will produce actual counts so we will need this functionality. We will probably just have to update the 'weight' name in this script. I dont think you need to change it now I'm just saying for info.

lizgzil replied:
I don't actually ever use the counts in the error metrics outputted other than to calculate the proportion of each feature category, so I'm not sure if it matters to even change the 'weight' name going forward? May have misinterpreted this though?

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+                orig_proportions, reweighted_proportions, target_proportions
+            )
+
+            prop_reweighted_counts_list = (
```
```suggestion
            prop_reweighted_counts_dict = (
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+                    prop_reweighted_counts_list["proportional_weight"],
+                )
+            )
+            prop_reweighted_proportions = calculate_proportions(prop_reweighted_counts)
```
I dont think you need to calculate proportions as they are already proportions. If this line is here because the proportions dont add up to 1, it would be because we have removed the dummy rows so you'll get it totalling <1 in some cases. 

This would actually change the weights in some cases.

lizgzil replied:
I think this was kept in the check that using either of 'weights' or 'proportional_weight' gave the same evaluation scores. But I think you are right and it is redundant to have it in - I will remove

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+                    reweighted_counts_list["weight"],
+                )
+            )
+            reweighted_proportions = calculate_proportions(reweighted_counts)
```
I think we need to join the dummy rows to the EPC dataset instead of drop them (what we've currently done) because we really need them for the 'reweighted' part of this analysis considering the way this script is written - specifically the `calculate_proportions` function.

This function calculates proportions based on ‘totals’, but these totals exclude the weights assigned to dummy rows. So for some of the LSOAs where dummy rows had to be added, we are actually changing the weights and then doing the error analysis on changed weights. This will have a bigger effect on LSOAs in EPC which were missing a category that has a high proportion in the target data. E.g. if a dummy row was given a fractional weight of 0.25 then this calculation in line 112 will calculate the proportions based on a total of `1 - 0.25`.

To fix this, we would have to do an outer join of the reweighted data onto EPC rather than left join and then re-run the reweighting pipeline and this error analysis.

Fixing this also means we will get a more accurate avg error reduction metric because the avg gets slightly skewed when we remove the dummy weights from the calculation.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    return args
+
+
+def get_errors_for_lsoa(
```
I think worth adding some notes within the function about what each section is doing :)

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+                target_proportions,
+            )
+
+        else:
+            # The weights may have been all Null
+            error_metrics_reweighted = {}
+            error_metrics_prop_reweighted = {}
+            average_error_reduction = None
+    else:
+        error_metrics_orig = {}
+        error_metrics_reweighted = {}
+        error_metrics_prop_reweighted = {}
+        average_error_reduction = None
+
+    final_metrics = {}
+    for metric in main_error_metrics:
+        final_metrics[metric] = {
+            "unweight": error_metrics_orig.get(metric, None),
+            "reweight": error_metrics_reweighted.get(metric, None),
+            "reweight-prop": error_metrics_prop_reweighted.get(metric, None),
+            "error_reduction": average_error_reduction,
+        }
+    return final_metrics
```
Just thinking about debugging this if we got an unexpected output later, you could have info logging statements if all weights are null or no target data is found and only return `final_metrics` dict if there actually are any metrics.

lizgzil replied:
good thinking

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    return final_metrics
+
+
+def load_epc_reweights(
```
You can put the cleaning function in `pipeline/enhance_epc/prepare_epc.py` :)

lizgzil replied:
good idea!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    return reweighted_epc_df
+
+
+def filter_epc(reweighted_epc_df: pl.DataFrame, target_features: dict) -> pl.DataFrame:
+    """
+    Filter the EPC dataset to just what is needed for calculating errors on.
+    This helps with processing time.
```
SMART

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    )
+    reweighted_epc_df = reweighted_epc_df.with_columns(
+        pl.Series(
+            name="nrooms", values=reweighted_epc_df["nrooms"].replace(None, "unknown")
```
What's the reason for this line im curious?

lizgzil replied:
I think because there was an 'unknown' category in the target dataset

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+
+    parser.add_argument(
+        "--sample",
+        help="Whether to test this script on a sample or not",
```
```suggestion
        help="Whether to test this script on a non-random sample or not",
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+    )
+
+    full_results = {}
+    for lsoa_code, epc_subset in tqdm(reweighted_epc_df_with_target.group_by("lsoa")):
```
oh cool I didnt know you could return a groupby into 2 variables this way!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,273 @@
+
+        # Save to S3
+
+    if not args.save_output:
+        name = args.reweighted_dir.replace("s3://asf-heat-pump-suitability/", "")
+        name = f"{name.split('.parquet')[0]}_evaluation.json"
+        if args.sample:
+            name = f"{name.split('.json')[0]}_sample.json"
+        args.save_output = name
```
nice!

crispy-wonton reviewed (commented):
Hi @lizgzil ,

Thanks so much for building these functions into a script and adding the s3 getters, really helpful!

I have left some comments, mostly on code documentation. The main important comments are those on `evaluate_reweighting.py` about the way proportions are calculated and used, given that we've removed dummy rows. I think the way the proportions for error analysis are currently calculated will actually change the weights for LSOAs where dummy rows were added. Please check my comments to see if you agree, and see the simple fix about joining weights to the EPC dataset (in the [reweighting pipeline](https://github.com/nestauk/asf_heat_pump_suitability/blob/dev/asf_heat_pump_suitability/pipeline/enhance_epc/run_script.py)) with an outer rather than left join.

Lastly, I didn't get a chance to test running the code yet but I wanted to get these comments through as soon as I discovered it. I will do that before this is merged.

Thanks so much! Let me know if you have questions :)

lizgzil committed:
Address Roisins PR comments - tidy docstrings, recalculate proportional error reduction, keep dummy rows in output of run_script, etc

lizgzil commented:
Thanks so much @crispy-wonton - your comments were extremely useful ! I've addressed your comments in this commit https://github.com/nestauk/asf_heat_pump_suitability/pull/28/commits/e75407168e71d3633c3ac0048929fed003571be0

I haven't rerun the run_script.py with the dummy rows included. I will try tomorrow and then rerun the evaluation.

lizgzil commented:
@crispy-wonton update:
- I have reran run_script.py with the outer join. There are 16788545 rows in this output. When the data is loaded we get this though:

![Screenshot 2024-07-12 at 09 58 30](https://github.com/user-attachments/assets/b063597e-2f2a-4788-9a56-c9f4dfba2645)

Which isn't what I expected (that row where UPRN= `10024356313.0` but UPRN_right = null?). I'm not quite sure if this matters, but it may represent a problem?

lizgzil commented:
latest evaluation is also run with changes to the script as in this PR + inclusion of dummy rows.
New results are better for property type than before.

The reweighting produces better Tenure proportions MAE 92.53% of the time
The reweighting produces the same Tenure proportions MAE 0.1% of the time
The reweighting produces better Property Type proportions MAE 93.5% of the time
The reweighting produces the same Property Type proportions MAE 0.17% of the time
The reweighting produces better Build Year proportions MAE 86.19% of the time
The reweighting produces the same Build Year proportions MAE 1.55% of the time

lizgzil committed:
Update evaluation notebook with changes

crispy-wonton commented:
> @crispy-wonton update:
> 
> * I have reran run_script.py with the outer join. There are 16788545 rows in this output. When the data is loaded we get this though:
> 
> ![Screenshot 2024-07-12 at 09 58 30](https://private-user-images.githubusercontent.com/15956065/348199400-b063597e-2f2a-4788-9a56-c9f4dfba2645.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NDU5MTQsIm5iZiI6MTcyMTY0NTYxNCwicGF0aCI6Ii8xNTk1NjA2NS8zNDgxOTk0MDAtYjA2MzU5N2UtMmYyYS00Nzg4LTlhNTYtYzlmNGRmYmEyNjQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDEwNTMzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0ZWUxMjMxMjIxMzg4NmFjOTU4MzdjMDljMjNjMjgyNTUwZjhkOGQ4YWI3ODUzZThiMWUxZDMzNTkyZDVmMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YDnjop8nvadEBOCQsUJlzSx6oxkAO8ggPRCTGP6AU4c)
> 
> Which isn't what I expected (that row where UPRN= `10024356313.0` but UPRN_right = null?). I'm not quite sure if this matters, but it may represent a problem?

I think this is due to the fact that in preparation for reweighting, we have to drop some EPC rows because of our reweighting exclusion criteria (any nulls in any reweighting feature ; or a category in EPC that isn't present in target data) or the LSOA is not present in the target data so the LSOA didnt get reweighted. In this case, I checked and can see that none of the properties in that LSOA received a weight. Therefore I think it's the latter reason although I didn't check the target data specifically to see if it was missing that LSOA.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/error_analysis/error_analysis_utils.py:
```diff
@@ -185,14 +128,105 @@ def process_single_lsoa(
+    target_proportions: dict,
+) -> np.float64:
+    """
+    Find the average relative error reduction for one feature (e.g. Tenure) and one LSOA before and after reweighting relative to before reweighting.
+
+    Args:
+        before_proportions (dict): The proportions of each category in the sample data before weighting, e.g. {'private': 0.672, 'rental': 0.328}
+        after_proportions (dict): The proportions of each category in the sample data after weighting.
+        target_proportions (dict): The proportions of each category in the target data.
+
+    Returns:
+        np.float64: The average relative error reduction before and after reweighting across each category for a feature in an LSOA.
+            This value is between -inf and 1
+
+        Interpreting the output:
+            - A negative value means weighting made the errors worse.
+            - A positive value means weighting improved the results.
+            - A very negative value (-inf) means weighting was much worse than no weighting.
+            - A value close to 1 means weighting made the results much better.
+            - A value of zero mean the errors with or without weighting were the same.
```
This docstring is perfect!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/error_analysis/error_analysis_utils.py:
```diff
@@ -185,14 +128,105 @@ def process_single_lsoa(
+            - A value of zero mean the errors with or without weighting were the same.
+    """
+
+    error_reduction_list = []
+    for key in before_proportions:
+        before_diff = np.abs(
+            before_proportions.get(key, 0) - target_proportions.get(key, 0)
+        )
+        after_diff = np.abs(
+            after_proportions.get(key, 0) - target_proportions.get(key, 0)
+        )
+        error_reduction = before_diff - after_diff
+        if before_diff == 0:
+            before_diff = 0.0000001  # To avoid division by 0
+        # The percentage reduction in error when switching between before and after weighting
+        error_reduction_list.append(error_reduction / before_diff)
+    if error_reduction_list:
+        return np.mean(error_reduction_list)
+    else:
+        return None
```
I think this is a good logic and works really well!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,271 @@
+                    feature_name (str): The feature name to calculate errors for
+                    lsoa_code (str): The LSOA
+                    target_features (dict): A nested dictionary of the counts for all property features and LSOAs
+                        For example, {'tenure': {'E01022833': {'owner-occupied': 548, 'rental (social)': 53}, 'E01013414': {}}, 'property_type': {}}
```
This example is really helpful! I think I should go back and add these type of things to the reweighting pipeline code that i wrote

lizgzil commented:
@crispy-wonton I just saw your comment - although I don't think you were concerned. So am I ok to merge this?

crispy-wonton commented on asf_heat_pump_suitability/pipeline/reweight_epc/evaluate_reweighting.py:
```diff
@@ -0,0 +1,271 @@
+    target_proportions = target_marginals[feature_name].get(lsoa_code)
+
+    if target_counts:
+        # If we have target information then calculate the errors
```
Just to check because I can't remember - will LSOAs with no target have an empty dict or will they be missing from `target_features` and `target_marginals` completely?

crispy-wonton reviewed (commented):
The updates look great - I think the error reduction calculation function is really clear now and all looks good. Thank you!

crispy-wonton commented:
> @crispy-wonton I just saw your comment - although I don't think you were concerned. So am I ok to merge this?

Yes I don't think it's a problem because I think it's an expected behaviour of the pipeline. I believe `UPRN` comes from the original EPC dataset and `UPRN_right` comes from the reweighted EPC rows. So any rows that were excluded from the reweighting because of our reweighting criteria would not be present in the reweighted EPC dataset, therefore they will appear as null UPRNs in the outer join.

I think this is good to merge :)
